package com.telefonia.embratel.Contato;

import javax.validation.ConstraintViolationException;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.test.context.junit4.SpringRunner;

import com.telefonia.embratel.Model.ModelJUnit;
import com.telefonia.embratel.Repository.RepositoryJUnit;

//@RunWith rodará com o spring runner,
//que suporta testes de integração de interfaces (testa se a 
//repository está funcionando)

@RunWith(SpringRunner.class)
@DataJpaTest //melhor testador de JPA
public class ContatoRepositoryIntegrationTest  {
private ModelJUnit Contato;
	
	@Autowired
	private RepositoryJUnit contatoRepository;
	
	@Before
	public void start() {
		Contato = new ModelJUnit("Gabriel", "011y", "9xxxxxxx9");
	}
	
	//regra, não aceita exceções, é a maneira de especificar
	//que a execução de um teste lançará uma exceção
	//Verifica se o @NotEmpty está funcionando
	@Rule
	public ExpectedException esperadaExcecao = ExpectedException.none();
	
	@Test
	public void salvarComTelNulo() throws Exception {
		esperadaExcecao.expect(ConstraintViolationException.class);
		esperadaExcecao.expectMessage("O Telefone deve ser preenchido");
		
		Contato.setTelefone(null);
		contatoRepository.save(Contato);
		
	}
	
	@Test
	public void salvarComDddNulo() throws Exception {
		esperadaExcecao.expect(ConstraintViolationException.class);
		esperadaExcecao.expectMessage("O DDD deve ser preenchido");
		
		Contato.setDdd(null);
		contatoRepository.save(Contato);
	}
	
	@Test
	public void salvarComNomeNulo() throws Exception {
		esperadaExcecao.expect(ConstraintViolationException.class);
		esperadaExcecao.expectMessage("O Nome deve ser preenchido");
		
		Contato.setNome(null);
		contatoRepository.save(Contato);
	}
	
}

